### DISCUSSION TITLE
Transfer-Encoding: gzip

### DISCUSSION THREAD

**Discussion by bagder** [hyperium/hyper#2633] [Q&A] [answered]
Is hyper doing anything special or assuming something special when `Transfer-Encoding: gzip` is used?

I'm trying to make curl [test 1122](https://github.com/curl/curl/blob/master/tests/data/test1122) to work, and in this I make hyper include the following two request headers:
~~~
Connection: TE
TE: gzip
~~~

... and then it gets a transfer-encoded response back, saying `Content-Length: 44` and then a gzip compressed body which happens to be exactly 44 bytes but it appears to me that hyper then doesn't consider that the end of the transfer.

Can you just clarify if this is known to be working fine with hyper?

**Comment by seanmonstar** [comment, marked as answer]
I just looked through the source code, and confirmed it is setting the body to "close-delimited". This is based on what it says in [RFC 7230 $3.3.3 ](https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.3):

> If a Transfer-Encoding header field is present in a response and the chunked transfer coding is not the final encoding, the message body length is determined by reading the connection until it is closed by the server.
>
> If a message is received with both a Transfer-Encoding and a Content-Length header field, the Transfer-Encoding overrides the Content-Length.

**Reply by bagder** [reply]
That text refers to *chunked* Transfer-Encoding, not to other transfer encodings. There's nothing wrong in sending gzip transfer-encoding with a content-length!

**Reply by seanmonstar** [reply]
Hm, if so, the RFC text _really_ makes that hard to understand. It specifically calls out a Transfer-Encoding without `chunked`. And then for if both exist, it only says "a Transfer-Encoding header field", not that that field ends with `chunked`. Is there anywhere else that clarifies it?

I started checking other HTTP libraries, and noticed a mix of handling. Some are extra strict (erroring if the `Transfer-Encoding` isn't exactly `chunked`), some of strict like hyper is, another defaults to strict like hyper with a "lenient" option, and another couple specifically only care if `chunked` was sent otherwise uses `Content-Length`. In a response, there's likely little damage that could be done, but that ambiguity in a request could lead to request smuggling.

**Reply by bagder** [reply]
With your existing logic there can't be *any* transfer-encoding that isn't also chunked (if we want persistent connections). I don't see how that makes any sense.

**Reply by bagder** [reply]
Then of course, using transfer-encoding for anything but chunked is very rarely used. And never by browsers for example.

**Reply by bagder** [reply]
But I will agree that the RFC says what you're saying. ~~Maybe I should take this to some HTTP peeps.~~

**Reply by bagder** [reply]
... the reason for this is that the `Content-Length` is the size of the **compressed** data, which means that when the transfer decoding changes the size, the amount may/will differ from the given size in the header.

Also detailed in [the Content-Length section](https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2).

**Reply by bagder** [reply]
I could perhaps add that when I added transfer-encoding support to curl, lots of years ago, the (live) server end we used provided content-lengths the way curl now supports. Now I need to ponder on how to proceed from here.

Thanks a lot @seanmonstar for your quick answers.

**Reply by bagder** [reply]
FYI: the next step went here: https://github.com/curl/curl/issues/7643
